### PR TITLE
Enforce required on template parameters in UI

### LINKED ISSUE #3984: Display/enforce required attribute on template parameters
Added in https://github.com/openshift/origin/pull/3963

Should at least indicate which fields are required, and ideally enforce a value being entered (for parameters without generators)

### PR DESCRIPTION
**Pull request by spadgett**
Fixes #3984

Edit mode:

openshift_web_console

Read only, no value:

openshift_web_console

### CODE REVIEW

**Comment by spadgett** [comment]
@liggitt ptal

File: assets/app/views/_templateopt.html [hunk 1]
```diff
@@ -1,4 +1,4 @@
-<div class="template-options" ng-show="hasParameters">
+<div class="template-options" ng-show="hasParameters" ng-form="paramForm">
```
Comment (liggitt) [inline assets/app/views/_templateopt.html]: do we ever include this twice in the same page?

Comment (spadgett) [inline assets/app/views/_templateopt.html]: The `template-options` directive? Not currently.

File: assets/app/views/_templateopt.html [hunk 2]
```diff
@@ -8,16 +8,37 @@
...
   </div>
-  <div class="form-group options" ng-repeat="parameter in template.parameters" ng-show="optionsExpanded">
-    <label for="param-{{$index}}" title="{{ parameter.description }}">{{parameter.name}}</label>
-    <input id="param-{{$index}}" class="form-control" type="text" placeholder="{{ parameter | parameterPlaceholder }}" ng-model="parameter.value">
+  <div
+      class="form-group options"
+      ng-repeat="parameter in template.parameters"
+      ng-show="optionsExpanded"
+      ng-init="paramID = 'param-' + $index">
+    <label ng-attr-for="{{paramID}}">{{parameter.name}}<span ng-if="parameter.required"> *</span></label>
```
Comment (liggitt) [inline assets/app/views/_templateopt.html]: do we have a class for the required indicator? if we don't, should we? we might want to make it red or something at some point...

File: assets/app/views/_templateopt.html (hunk 1 above)
Comment (liggitt) [inline assets/app/views/_templateopt.html]: if we did, would a fixed name break?

File: assets/app/views/_templateopt.html (hunk 2 above)
Comment (liggitt) [inline assets/app/views/_templateopt.html]: @jwforres what's your feeling on leading vs trailing required indicators? I think I prefer leading with negative margin to allow the labels to continue to left-align. That lets the indicators align, and still appear if the labels get truncated.

**Comment by liggitt** [comment]
nits, LGTM. Eventually, we'll want to parse error responses from the template processing, correlate errors with parameters, and display the returned errors in the form. But not today.

File: assets/app/views/_templateopt.html (hunk 1 above)
Comment (spadgett) [inline assets/app/views/_templateopt.html]: I don't think it's a problem from my reading of the documentation for `ngForm`:

https://docs.angularjs.org/api/ng/directive/form

File: assets/app/views/_templateopt.html (hunk 2 above)
Comment (jwforres) [inline assets/app/views/_templateopt.html]: I would prefer leading indicators

Comment (spadgett) [inline assets/app/views/_templateopt.html]: I'll add a style that handles it.

File: assets/app/views/_templateopt.html (hunk 1 above)
Comment (spadgett) [inline assets/app/views/_templateopt.html]: We would have a problem with duplicate IDs, but that's not specific to this pull.

**Comment by jwforres** [comment]
Something we did on another project was to have a checkbox to show only the required fields, some of these templates have a LOT of params.  Thoughts?   (can be a follow-up issue)

File: assets/app/views/_templateopt.html (hunk 2 above)
Comment (spadgett) [inline assets/app/views/_templateopt.html]: Ugh, negative margin doesn't work with labels that use class `truncate`. The asterisk is hidden.

Comment (spadgett) [inline assets/app/views/_templateopt.html]: Read-only

openshift_web_console

Edit mode:

openshift_web_console

Comment (spadgett) [inline assets/app/views/_templateopt.html]: New project form:

openshift_web_console

**Comment by spadgett** [comment]
We could also have field sets for required and optional, required at the top.

File: assets/app/views/_templateopt.html (hunk 2 above)
Comment (liggitt) [inline assets/app/views/_templateopt.html]: I think that looks great

**Comment by spadgett** [comment]
@liggitt Pushed the asterisk change.

**Comment by liggitt** [comment]
Let's not re-order fields yet... there's mutterings in https://github.com/openshift/origin/issues/3973 about field grouping, etc

File: assets/app/styles/_core.less [hunk 3]
```diff
@@ -301,6 +301,14 @@ html, body {
   }
 }
 
+label.required:before {
+  content: '* ';
+  // Use position: absolute instead of negative margin so the asterisk shows
+  // up when overflow is hidden.
+  position: absolute;
+  left: 10px;
```
Comment (liggitt) [inline assets/app/styles/_core.less]: doesn't this inset 10px from the left? does this presume there's at least 10px+width('*') of left padding?

Comment (spadgett) [inline assets/app/styles/_core.less]: You're right, this is wrong. It just happened to work :(

I'm not sure how to fix the problem with `overflow: hidden;`

**Comment by spadgett** [comment]
I finally have a fix. Here is the asterisk at mobile resolution with overflow: hidden.

openshift_web_console_and_1__sam_dhcp137-192____go_src_github_com_openshift_origin__zsh_

**Comment by liggitt** [comment]
glad to see the asterisk working, though now I'm going to ask why the labels truncate at 50% of the available width... shouldn't the value stack vertically so the label has a chance of being shown? fine in a follow-up, but let's fix that

File: assets/app/styles/_core.less [hunk 4]
```diff
@@ -301,6 +314,11 @@ html, body {
   }
 }
 
+.required:before {
+  content: '* ';
```
Comment (liggitt) [inline assets/app/styles/_core.less]: probably don't need the trailing space, since only left-alignment matters

Comment (liggitt) [inline assets/app/styles/_core.less]: is `.required` too general a class-name? do you want to scope this to `label.required`?

Comment (spadgett) [inline assets/app/styles/_core.less]: That's what I had before, but I took it out experimenting with spans.... :-/

I'll fix both issues.

**Comment by spadgett** [comment]
The stacking behavior is inconsistent in edit mode vs. read only mode as well. (Edit always stacks.) The input fields are also *wider* at mobile resolutions than at full screen widths, which is strange.

File: assets/app/styles/_core.less (hunk 4 above)
Comment (spadgett) [inline assets/app/styles/_core.less]: The space matters. Without it:

openshift_web_console

**Comment by liggitt** [comment]
LGTM

**Comment by jwforres** [comment]
The difference in whitespace between the label (or input), the description, and the error message is not consistent, can we clean that up easily?  Otherwise seems fine.

**Comment by spadgett** [comment]
Cleaned up the margins.

openshift_web_console

openshift_web_console

**Comment by liggitt** [comment]
[merge]

**Pull request merged**